chainer/variable.py

```diff
@@ -284,6 +284,20 @@ def add_cand(cand):
             for x, gx in zip(func.inputs, gxs):
                 if gx is None:
                     continue
+
+                if not isinstance(gx, type(x.data)):
+                    raise TypeError(
+                        'Type of data and grad mismatch in %s:\n%s != %s'
+                        % (func.label, type(x.data), type(gx)))
+                if gx.dtype != x.data.dtype:
+                    raise TypeError(
+                        'Dtype of data and grad mismatch in %s:\n%s != %s'
+                        % (func.label, x.data.dtype, gx.dtype))
+                if gx.shape != x.data.shape:
+                    raise ValueError(
+                        'Shape of data and grad mismatch in %s:\n%s != %s'
+                        % (func.label, x.data.shape, gx.shape))
```

Comment:
Could you unify the type checking code here with that in the grad.setter? It seems redundant (except showing the function label). https://github.com/pfnet/chainer/blob/master/chainer/variable.py#L153